Date: Tue, 2 Sep 2025 02:29:33 +0200
Subject: [PATCH] cdc-ncm: fix linking with generic CarPlay Headunits
Patch tested on:

Product: MB Infotainment
Manufacturer: Daimler AG
idVendor: 0x25e1
idProduct: 0x4351
Upstream-Status: Pending
---
 drivers/net/usb/cdc_ncm.c   | 83 ++++++++++++++++++++++++++++++-------
 drivers/net/usb/usbnet.c    |  2 +-
 include/linux/usb/cdc_ncm.h | 10 +++--
 3 files changed, 76 insertions(+), 19 deletions(-)

diff --git a/drivers/net/usb/cdc_ncm.c b/drivers/net/usb/cdc_ncm.c
index 4abfdfcf0e28..eb5df32ab8ed 100644
--- a/drivers/net/usb/cdc_ncm.c
+++ b/drivers/net/usb/cdc_ncm.c
@@ -711,6 +711,13 @@ static int cdc_ncm_setup(struct usbnet *dev)
 	def_tx = min_t(u32, CDC_NCM_NTB_DEF_SIZE_TX,
 		       le32_to_cpu(ctx->ncm_parm.dwNtbOutMaxSize));
 
+	if (dev->driver_info->flags & CDC_NCM_FLAG_CARPLAY_MAXIMIZE_RX_TX) {
+		def_rx = min_t(u32, CDC_NCM_NTB_MAX_SIZE_RX,
+			le32_to_cpu(ctx->ncm_parm.dwNtbInMaxSize));
+		def_tx = min_t(u32, CDC_NCM_NTB_MAX_SIZE_TX,
+			le32_to_cpu(ctx->ncm_parm.dwNtbOutMaxSize));
+	}
+
 	/* clamp rx_max and tx_max and inform device */
 	cdc_ncm_update_rxtx_max(dev, def_rx, def_tx);
 
@@ -902,21 +909,22 @@ int cdc_ncm_bind_common(struct usbnet *dev, struct usb_interface *intf, u8 data_
 	/* Device-specific flags */
 	ctx->drvflags = drvflags;
 
-	/* Reset data interface. Some devices will not reset properly
-	 * unless they are configured first.  Toggle the altsetting to
-	 * force a reset.
-	 * Some other devices do not work properly with this procedure
-	 * that can be avoided using quirk CDC_MBIM_FLAG_AVOID_ALTSETTING_TOGGLE
-	 */
-	if (!(ctx->drvflags & CDC_MBIM_FLAG_AVOID_ALTSETTING_TOGGLE))
-		usb_set_interface(dev->udev, iface_no, data_altsetting);
-
-	temp = usb_set_interface(dev->udev, iface_no, 0);
-	if (temp) {
-		dev_dbg(&intf->dev, "set interface failed\n");
-		goto error2;
+	if (!(dev->driver_info->flags & CDC_NCM_FLAG_CARPLAY_ALWAYS_UP)) {
+		/* Reset data interface. Some devices will not reset properly
+		* unless they are configured first.  Toggle the altsetting to
+		* force a reset.
+		* Some other devices do not work properly with this procedure
+		* that can be avoided using quirk CDC_MBIM_FLAG_AVOID_ALTSETTING_TOGGLE
+		*/
+		if (!(ctx->drvflags & CDC_MBIM_FLAG_AVOID_ALTSETTING_TOGGLE))
+			usb_set_interface(dev->udev, iface_no, data_altsetting);
+
+		temp = usb_set_interface(dev->udev, iface_no, 0);
+		if (temp) {
+			dev_dbg(&intf->dev, "set interface failed\n");
+			goto error2;
+		}
 	}
-
 	/* initialize basic device settings */
 	if (cdc_ncm_init(dev))
 		goto error2;
@@ -981,6 +989,15 @@ int cdc_ncm_bind_common(struct usbnet *dev, struct usb_interface *intf, u8 data_
 	dev->net->netdev_ops = &cdc_ncm_netdev_ops;
 	dev->net->max_mtu = cdc_ncm_max_dgram_size(dev) - cdc_ncm_eth_hlen(dev);
 
+	if (dev->driver_info->flags & CDC_NCM_FLAG_CARPLAY_MAXIMIZE_MTU) {
+		dev->net->mtu = dev->net->max_mtu;
+	}
+
+	if (dev->driver_info->flags & CDC_NCM_FLAG_CARPLAY_ALWAYS_UP) {
+		if (!netif_carrier_ok(dev->net))
+			netif_carrier_on(dev->net);
+	}
+
 	return 0;
 
 error2:
@@ -1882,11 +1899,19 @@ static void cdc_ncm_status(struct usbnet *dev, struct urb *urb)
 		/* RTL8156 shipped before 2021 sends notification about
 		 * every 32ms. Don't forward notification if state is same.
 		 */
+		if (dev->driver_info->flags & CDC_NCM_FLAG_CARPLAY_ALWAYS_UP) {
+			dev_info(&dev->udev->dev,
+				"NCM: ignoring USB_CDC_NOTIFY_NETWORK_CONNECTION, always UP\n");
+			return;
+		}
+
 		if (netif_carrier_ok(dev->net) != !!event->wValue)
 			usbnet_link_change(dev, !!event->wValue, 0);
 		break;
 
 	case USB_CDC_NOTIFY_SPEED_CHANGE:
+		dev_info(&dev->udev->dev,
+			"NCM: received USB_CDC_NOTIFY_SPEED_CHANGE\n");
 		if (urb->actual_length < (sizeof(*event) +
 					sizeof(struct usb_cdc_speed_change)))
 			set_bit(EVENT_STS_SPLIT, &dev->flags);
@@ -1907,10 +1932,38 @@ static void cdc_ncm_update_filter(struct usbnet *dev)
 {
 	struct cdc_ncm_ctx *ctx = (struct cdc_ncm_ctx *)dev->data[0];
 
+	if (ctx->filtering_supported && (dev->driver_info->flags & CDC_NCM_FLAG_CARPLAY_MATCH_FILTER)) {
+		ctx->filtering_supported = false;
+		usb_control_msg(dev->udev,
+			usb_sndctrlpipe(dev->udev, 0),
+			USB_CDC_SET_ETHERNET_PACKET_FILTER,
+			USB_TYPE_CLASS | USB_RECIP_INTERFACE,
+			0x1e /* cdc_filter */,
+			dev->intf->cur_altsetting->desc.bInterfaceNumber,
+			NULL,
+			0,
+			USB_CTRL_SET_TIMEOUT
+		);
+	}
+
 	if (ctx->filtering_supported)
 		usbnet_cdc_update_filter(dev);
 }
 
+static const struct driver_info carplay_info = {
+	.description = "Generic Apple CarPlay Headunit (MFI specification compatible)",
+	.flags = FLAG_POINTTOPOINT | FLAG_NO_SETINT | FLAG_MULTI_PACKET
+			| FLAG_ETHER | FLAG_SEND_ZLP | CDC_NCM_FLAG_CARPLAY_MATCH_FILTER | CDC_NCM_FLAG_CARPLAY_ALWAYS_UP 
+			| CDC_NCM_FLAG_CARPLAY_MAXIMIZE_RX_TX | CDC_NCM_FLAG_CARPLAY_MAXIMIZE_MTU,
+	    .bind = cdc_ncm_bind,
+	    .unbind = cdc_ncm_unbind,
+	    .manage_power = usbnet_manage_power,
+	    .status = cdc_ncm_status,
+	    .rx_fixup = cdc_ncm_rx_fixup,
+	    .tx_fixup = cdc_ncm_tx_fixup,
+	    .set_rx_mode = cdc_ncm_update_filter,
+};
+
 static const struct driver_info cdc_ncm_info = {
 	.description = "CDC NCM (NO ZLP)",
 	.flags = FLAG_POINTTOPOINT | FLAG_NO_SETINT | FLAG_MULTI_PACKET
@@ -2101,7 +2154,7 @@ static const struct usb_device_id cdc_devs[] = {
 	/* Generic CDC-NCM devices */
 	{ USB_INTERFACE_INFO(USB_CLASS_COMM,
 		USB_CDC_SUBCLASS_NCM, USB_CDC_PROTO_NONE),
-		.driver_info = (unsigned long)&cdc_ncm_info,
+		.driver_info = (unsigned long)&carplay_info,
 	},
 	{
 	},
diff --git a/drivers/net/usb/usbnet.c b/drivers/net/usb/usbnet.c
index ccf45ca2feb5..0a81eb9ebe22 100644
--- a/drivers/net/usb/usbnet.c
+++ b/drivers/net/usb/usbnet.c
@@ -44,7 +44,7 @@
  * more before an irq is required, under load.  Jumbograms change
  * the equation.
  */
-#define	MAX_QUEUE_MEMORY	(60 * 1518)
+#define	MAX_QUEUE_MEMORY	(512 * 1518)
 #define	RX_QLEN(dev)		((dev)->rx_qlen)
 #define	TX_QLEN(dev)		((dev)->tx_qlen)
 
diff --git a/include/linux/usb/cdc_ncm.h b/include/linux/usb/cdc_ncm.h
index 4ac082a63173..a0dda6028bad 100644
--- a/include/linux/usb/cdc_ncm.h
+++ b/include/linux/usb/cdc_ncm.h
@@ -85,9 +85,13 @@
 #define CDC_NCM_TIMER_INTERVAL_MAX		(U32_MAX / NSEC_PER_USEC)
 
 /* Driver flags */
-#define CDC_NCM_FLAG_NDP_TO_END			0x02	/* NDP is placed at end of frame */
-#define CDC_MBIM_FLAG_AVOID_ALTSETTING_TOGGLE	0x04	/* Avoid altsetting toggle during init */
-#define CDC_NCM_FLAG_PREFER_NTB32 0x08	/* prefer NDP32 over NDP16 */
+#define CDC_NCM_FLAG_NDP_TO_END			1 << 1	/* NDP is placed at end of frame */
+#define CDC_MBIM_FLAG_AVOID_ALTSETTING_TOGGLE	1 << 2	/* Avoid altsetting toggle during init */
+#define CDC_NCM_FLAG_PREFER_NTB32 1 << 3	/* prefer NDP32 over NDP16 */
+#define CDC_NCM_FLAG_CARPLAY_MATCH_FILTER 1 << 16
+#define CDC_NCM_FLAG_CARPLAY_ALWAYS_UP 1 << 17
+#define CDC_NCM_FLAG_CARPLAY_MAXIMIZE_RX_TX 1 << 18
+#define CDC_NCM_FLAG_CARPLAY_MAXIMIZE_MTU 1 << 19
 
 #define cdc_ncm_comm_intf_is_mbim(x)  ((x)->desc.bInterfaceSubClass == USB_CDC_SUBCLASS_MBIM && \
 				       (x)->desc.bInterfaceProtocol == USB_CDC_PROTO_NONE)
-- 
2.52.0

